File: app/models/conversations/group_conversation.rb
# frozen_string_literal: true

module Conversations
  class GroupConversation < Conversation
    scope :groups, GroupConversationsQuery

    # Virtual attributes for query results - using attribute method for proper mapping
    attribute :member_count, :integer
    attribute :has_unread_messages, :boolean
    attribute :is_member, :boolean
    attribute :description, :string

    validates :name, presence: true, length: { minimum: 1, maximum: 100 }

    # Manual attribute assignment for query results (fallback if attribute method doesn't work)
    def assign_query_attributes(attrs = {})
      self.member_count = attrs['member_count'] if attrs.key?('member_count')
      self.has_unread_messages = attrs['has_unread_messages'] if attrs.key?('has_unread_messages') 
      self.is_member = attrs['is_member'] if attrs.key?('is_member')
      self.description = attrs['description'] if attrs.key?('description')
      self
    end

    def group?
      true
    end

    def direct?
      false
    end

    def as_json
      {
        id:,
        name:,
        description:,
        is_public:,
        created_by_id:,
        created_at: created_at.iso8601,
        member_count:,
        has_unread_messages:,
        is_member:
      }
    end
  end
end

File: frontend/src/models/groupConversation.ts
import { toUser } from './user';
import type { APIUser, User } from './user';

export interface GroupConversation {
  id: number;
  name: string;
  description: string | null;
  isPublic: boolean;
  createdById: number;
  createdAt: Date;
  memberCount: number;
  hasUnreadMessages: boolean;
  isMember: boolean;
}

export interface APIGroupConversation {
  id: number;
  name: string;
  description: string | null;
  is_public: boolean;
  created_by_id: number;
  created_at: string;
  member_count: number;
  has_unread_messages: boolean;
  is_member: boolean;
}

export interface DirectConversation {
  id: number;
  createdById: number;
  createdAt: Date;
  hasUnreadMessages: boolean;
  participants: User[];
}

export interface APIDirectConversation {
  id: number;
  created_by_id: number;
  created_at: string;
  has_unread_messages: boolean;
  participants: APIUser[];
}

export function toGroupConversation(data: APIGroupConversation): GroupConversation {
  return {
    id: data.id,
    name: data.name,
    description: data.description,
    isPublic: data.is_public,
    createdById: data.created_by_id,
    createdAt: new Date(data.created_at),
    memberCount: data.member_count,
    hasUnreadMessages: data.has_unread_messages,
    isMember: data.is_member,
  };
}

export function toDirectConversation(data: APIDirectConversation): DirectConversation {
  return {
    id: data.id,
    createdById: data.created_by_id,
    createdAt: new Date(data.created_at),
    hasUnreadMessages: data.has_unread_messages,
    participants: data.participants.map(toUser),
  };
}
